plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPropertyPage.java

```diff
@@ -105,109 +116,93 @@ public void run() {
+    } else if (facetedProject != null && AppEngineFlexFacet.hasAppEngineFacet(facetedProject)) {
+      createFlexPanelIfNeeded();
+      stackLayout.topControl = flexPreferencesPanel;
+      databindingSupport =
+          PreferencePageSupport.create(this, flexPreferencesPanel.getDataBindingContext());
     } else {
-      logger.log(Level.WARNING,
-          "App Engine Deployment property page is only visible if project contains an App Engine facet");
-      return null;
+      createBlankPanelIfNeeded();
+      stackLayout.topControl = blankPreferencesPanel;
+      databindingSupport =
+          PreferencePageSupport.create(this, blankPreferencesPanel.getDataBindingContext());
     }
-  }
+    container.layout();
 
-  @Override
-  public void setVisible(boolean visible) {
-    super.setVisible(visible);
-    evaluateFacetConfiguration();
   }
 
-  /**
-   * Checks to see if the project associated with this Property dialog still has a valid
-   * App Engine facet. If it does, allow messages to be displayed. If it doesn't display
-   * appropriate error message and prevent other messages from being displayed.
-   */
-  private void evaluateFacetConfiguration() {
-    if (isStandardPanel && !AppEngineStandardFacet.hasAppEngineFacet(facetedProject)) {
-      IProjectFacet projectFacet = ProjectFacetsManager.getProjectFacet(AppEngineStandardFacet.ID);
-      invalidFacetConfigErrorMessage =
-          Messages.getString("invalid.deploy.page.state", projectFacet.getLabel());
-      updatePageBasedOnFacetConfig(false);
-    } else if (!isStandardPanel && !AppEngineFlexFacet.hasAppEngineFacet(facetedProject)) {
-      IProjectFacet projectFacet = ProjectFacetsManager.getProjectFacet(AppEngineFlexFacet.ID);
-      invalidFacetConfigErrorMessage =
-          Messages.getString("invalid.deploy.page.state", projectFacet.getLabel());
-      updatePageBasedOnFacetConfig(false);
-    } else {
-      invalidFacetConfigErrorMessage = "";
-      updatePageBasedOnFacetConfig(true);
+  private void createBlankPanelIfNeeded() {
+    if (blankPreferencesPanel == null) {
+      blankPreferencesPanel = new BlankDeployPreferencesPanel(container);
+      blankPreferencesPanel.setFacetPageSelector(new Runnable() {
+        @Override
+        public void run() {
+          PreferenceDialog preferenceDialog = (PreferenceDialog)getContainer();
+          TreeItem[] items = preferenceDialog.getTreeViewer().getTree().getItems();
+          for (TreeItem treeItem : items) {
+            if (FACET_PAGE_ID.equals(((PreferenceNode)treeItem.getData()).getId())) {
+              preferenceDialog.getTreeViewer().setSelection(
+                  new StructuredSelection(treeItem.getData()), true);
+            }
+          }
+        }
+      });
     }
+  }
 
+  private void createStandardPanelIfNeeded() {
+    if (standardPreferencesPanel == null) {
+      IGoogleLoginService loginService =
+          PlatformUI.getWorkbench().getService(IGoogleLoginService.class);
```

Comment:
If you like, you could use `ServiceContextFactory` to create this class and `@Inject` these values.